### PR TITLE
rustup Spec

### PR DESCRIPTION
**Pull request by geenva**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
New specification for `rustup`.

**What is the current behavior? (You can also link to an open issue here)**
Completed. Awaiting merge.

**What is the new behavior (if this is a feature change)?**
There is an additional `rustup` completion for Fig.

~~**Additional info:**~~

### CODE REVIEW

**Comment by geenva** [comment]
Hi, sorry for abandoning my PR earlier. I will now continue to work on it.

**Comment by geenva** [comment]
Ready to review! Thanks.

**Review by Milo123459 — APPROVED** [review]
I don't see any problems. LGTM

File: src/rustup.ts [hunk 1]
```diff
@@ -0,0 +1,711 @@
+// To learn more about Fig's autocomplete standard visit: https://fig.io/docs/concepts/cli-skeleton
+
+// The below is a dummy example for git. Make sure to change the file name!
```
Comment (grant0417) [inline src/rustup.ts:3]: Remove these comments or add more specific ones for rustup

File: src/rustup.ts [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+        },
+        {
+          name: "help",
+          description:
+            "Prints this message or the help of the given subcommand(s)",
+        },
+      ],
+
+      // If a subcommand or option takes an argument, you must include the args prop, even if it's an empty object (like below)
+      // If you want to build custom suggestions for arguments check out: https://fig.io/docs/concepts/dynamic-suggestions
```
Comment (QuiiBz) [inline src/rustup.ts]: Is this comment still needed?

File: src/rustup.ts [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+        {
+          name: ["-h", "--help"],
+          description: "Prints help information",
+        },
+      ],
+    },
+    {
+      name: "update",
+      description: "Update Rust toolchains and rustup",
+
```
Comment (QuiiBz) [inline src/rustup.ts:41]: Extra line

**Review by grant0417 — CHANGES_REQUESTED** [review]
Looks good to me, only issue is the template comments.

File: src/rustup.ts [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+            },
+            {
+              name: ["-V", "--version"],
+              description: "Prints version information",
+            },
+          ],
+        },
+        {
+          name: "add",
+          description: "Add a target to a Rust toolchain",
```
Comment (QuiiBz) [inline src/rustup.ts:122]: Doesn’t this sub command have an argument, which is the target to add to the rust tool chain?

File: src/rustup.ts [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+            {
+              name: ["-h", "--help"],
+              description: "Prints help information",
+            },
+            {
+              name: ["-V", "--version"],
+              description: "Prints version information",
+            },
+          ],
+          description: "Remove a target from a Rust toolchain",
```
Comment (QuiiBz) [inline src/rustup.ts:146]: Same as above

File: src/rustup.ts [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+            {
+              name: ["-h", "--help"],
+              description: "Prints help information",
+            },
+            {
+              name: ["-V", "--version"],
+              description: "Prints version information",
+            },
+          ],
+          description: "Install or update a given toolchain",
```
Comment (QuiiBz) [inline src/rustup.ts:205]: Same as above, looks like this issue is repeating a few times underneath

**Review by grant0417 — APPROVED** [review]

File: src/rustup.ts [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+      ],
+      args: [
+        {
+          name: "<toolchain>",
+          description:
+            "Toolchain name, such as 'stable', 'nightly', or '1.8.0'",
+        },
+        {
+          name: "<command>",
+          isVariadic: true,
```
Comment (QuiiBz) [inline src/rustup.ts:404]: Should we add the `isCommand` flag? See https://fig.io/docs/reference/arg#iscommand

File: src/rustup.ts [hunk 8]
```diff
@@ -0,0 +1,711 @@
...
+          name: ["-h", "--help"],
+          description: "Prints help information",
+        },
+        {
+          name: "--nomicon",
+          description: "The Dark Arts of Advanced and Unsafe Rust Programming",
+        },
+        {
+          name: "--path",
+          description: "Only print the path tp the documentation",
```
Comment (QuiiBz) [inline src/rustup.ts:466]: Typo:
```suggestion
          description: "Only print the path to the documentation",
```

File: src/rustup.ts [hunk 9]
```diff
@@ -0,0 +1,711 @@
...
+            "Toolchain name, such as 'stable', 'nightly', or '1.8.0'",
+          args: {
+            name: "<toolchain>",
+          },
+        },
+      ],
+      args: {
+        name: "<topic>",
+        description:
+          "Topic such as 'core', 'fn', 'usize', 'eprintln!', 'core::arch', 'alloc::format!', 'std::fs', 'std::fs::read_dir', 'std::io::Bytes', 'std::iter::Sum', 'std::io::error::Result' etc",
```
Comment (QuiiBz) [inline src/rustup.ts:515]: Could you add these suggestions to the [suggestions](https://fig.io/docs/reference/arg#suggestions) property?

**Review by QuiiBz — CHANGES_REQUESTED** [review]
Hi, thanks for creating this PR!

File: src/rustup.ts (hunk 9 above)
Comment (grant0417) [reply src/rustup.ts:515]: This could be any topic in the lang, i don't know if these would be good as suggestions. A generator would probably be needed.

Comment (QuiiBz) [reply src/rustup.ts:515]: Thanks for the clarification, would it be possible to get a list of them with a shell command, so we could make a generator? If not, that’s 100% fine to leave it like this!

Comment (grant0417) [reply src/rustup.ts:515]: Here is a oneliner to grab all the pages in std.
`find $(rustup docs --path | sed -e "s|index\.html|std|") | grep "\.html" | sed -E -e "s|^(.*)/html/std|std|" -e "s|\.html||" -e "s|/|::|g" -e "s/constant\.|trait\.|struct\.|macro\.|fn\.|keyword\.|primitive\.|type\.|enum\.|union\.|::index//" | grep std`

**Comment by geenva** [comment]
Think everything is alright now?

File: src/rustup.ts (hunk 9 above)
Comment (geenva) [reply src/rustup.ts:580]: > Here is a oneliner to grab all the pages in std. `find $(rustup docs --path | sed -e "s|index.html|std|") | grep "\.html" | sed -E -e "s|^(.*)/html/std|std|" -e "s|\.html||" -e "s|/|::|g" -e "s/constant\.|trait\.|struct\.|macro\.|fn\.|keyword\.|primitive\.|type\.|enum\.|union\.|::index//"`

Thanks so much for this, helped me a lot!

File: src/rustup.ts [hunk 10]
```diff
@@ -0,0 +1,2458 @@
...
+        suggestions: [
```
Comment (grant0417) [inline src/rustup.ts:581]: You can actually generate those all at run time which will always be up to date using a generator

```suggestion
        generators: {
          script: `find $(rustup docs --path | sed -e "s|index\\.html|std|") $(rustup docs --path | sed -e "s|index\.html|alloc|") $(rustup docs --path | sed -e "s|index\\.html|core|") | grep "\\.html" | sed -E -e "s|^(.*)/html/||" -e "s|\\.html||" -e "s|/|::|g" -e "s/constant\\.|trait\\.|struct\\.|macro\\.|fn\\.|keyword\\.|primitive\\.|type\\.|enum\\.|union\\.|traitalias\\.|::index$//"`,
          splitOn: "\n"
        },
```

**Review by grant0417 — CHANGES_REQUESTED** [review]

File: src/rustup.ts (hunk 10 above)
Comment (geenva) [reply src/rustup.ts:581]: Whoa, didn’t know that. Thanks!

**Comment by geenva** [comment]
Thanks for checking this PR so thoroughly!

**Comment by grant0417** [comment]
Almost there, you just forgot to remove all of the suggestion lines.

**Comment by geenva** [comment]
> Almost there, you just forgot to remove all of the suggestion lines.

Of course, sorry about that. Fixed.

**Review by grant0417 — APPROVED** [review]
Awesome job!

**Comment by mschrage** [comment]
Let's get this merged in! ✨

**Comment by geenva** [comment]
Uh oh, I will lint it on my end and recommit.

**Review by mschrage — APPROVED** [review]
🤩

**Pull request merged**